state/azure/cosmosdb/storedprocedure.go

```diff
@@ -0,0 +1,118 @@
+// ------------------------------------------------------------
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+// ------------------------------------------------------------
+
+package cosmosdb
+
+const spDefinition string = `// upserts - an array of objects to upsert
+// deletes - an array of objects to delete
+
+function sample(upserts, deletes) {
+    var context = getContext();
+    var container = context.getCollection();
+    var response = context.getResponse();
+    
+    if (typeof upserts === "string") {
+        throw new Error("first arg is a string, expected array of objects");
+    }
+
+    if (typeof deletes === "string") {
+        throw new Error("second arg is a string, expected array of objects");
+    }
+
+    // create the query string used to look up deletes
+    var query = "select * from n where n.id = ";
+    if (deletes.length > 0) {
+        query += ("'" + deletes[0].id + "'");
+
+        for (let j = 1; j < deletes.length; j++) {
+            query += (" or n.id = '" + deletes[j].id + "'");
+        }
+    }
+    
+    var upsertCount = 0;
+    var deleteCount = 0;
+      
+    var collectionLink = container.getSelfLink();
+
+    // do the upserts first    
+    if (upserts.length != 0) {
+        tryCreate(upserts[upsertCount], callback);
+    } else {
+        tryQueryAndDelete();
+    }
+
+    function tryCreate(doc, callback) {        
+        var isAccepted = container.upsertDocument(collectionLink, doc, callback);
+        
+        // fail if we hit execution bounds
+        if (!isAccepted) {                        
+            throw new Error("upsertDocument() not accepted, please retry");
+        }
+    }
+
+    function callback(err, doc, options) {        
+        if (err) throw err;
+
+        upsertCount++;
+
+        if (upsertCount >= upserts.length) {
+            
+            // upserts are done, start the deletes, if any
+            if (deletes.length > 0) {
+                tryQueryAndDelete()
+            }
+        } else {            
+            tryCreate(upserts[upsertCount], callback);
+        }
+    }
+
+    function tryQueryAndDelete(continuation) {        
+        var requestOptions = { continuation: continuation };
+        
+        var isAccepted = container.queryDocuments(collectionLink, query, requestOptions, function (err, retrievedDocs, responseOptions) {
+            if (err) {
+                throw err;
+            }
+
+            if (retrievedDocs == null) {                
+                response.setBody(JSON.stringify("success"));
+            } else if (retrievedDocs.length > 0) {                
+                tryDelete(retrievedDocs);
+            } else if (responseOptions.continuation) {                
```

Comment:
look like this code is for the pagination. so that this needs to be called even if retrievedDocs.length > 0.

Can you please double-check the behavior of this code?

Reply:
Removed the pagination code.  We actually don't want pagination/continuation because then it will require multiple sp invocations and it won't be atomic.

Reply:
What’s the max items per page ?

Reply:
It depends on configuration which dapr won't have direct control over.  The user will have to configure this when they create the container to suit their needs -  https://docs.microsoft.com/en-us/azure/cosmos-db/request-units